### PR TITLE
Metrics in node level

### PR DESCRIPTION
**Pull request by jngz-es**
*Issue #23 , if available:*

*Description of changes:*
Add metrics in SQL plugin.
Currently ignore metrics integration test case, because
1) ES integration testing framework doesn't support singleton/static using.
2) nodeSettings can't add parameters into the initial settings of ES cluster.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+public interface Counter<T> {
+
+    void increment();
+
+    void increment(long n);
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: Should we stick to java's naming convention (see `LongAdder::add(long)`) and call this method `add` instead?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import java.util.function.Supplier;
+
+public class GaugeMetric<T> extends Metric<T> {
+
+    private Supplier<T> LOAD_VALUE;
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: why is the field name in all caps?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import java.util.function.Supplier;
+
+public class GaugeMetric<T> extends Metric<T> {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: can you add documentation for all classes (at least non-trivial ones) to make the purpose of the class clear.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    REQ_TOTAL("request_total", 0),
+    REQ_COUNT_TOTAL("request_count", 1),
+    FAILED_REQ_COUNT_SYS("failed_request_count_syserr", 1),
+    FAILED_REQ_COUNT_CUS("failed_request_count_cuserr", 1),
+    FAILED_REQ_COUNT_CB("failed_request_count_cb", 1),
+    CIRCUIT_BREAKER("circuit_breaker", 2),
+    DEFAULT("default", 0);
+
+    private String name;
+    private int type;
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: np: `typeId` ?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    public int getType() {
+        return type;
+    }
+
+    public static List<String> getNames() {
+        return Arrays.stream(MetricType.values()).map(v -> v.name).collect(Collectors.toList());
+    }
+
+    public boolean isNumerical() {
+        return type < 2;
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: This looks really neat, but I think it's also not very extensible. First, it requires maintaining the order of the values, and second, if one adds a new value, s/he might have to update this number as well, so it's extra maintenance, plus implicit requirement. I suggest we explicitly whitelist the types here.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import com.amazon.opendistroforelasticsearch.sql.query.join.BackOffRetryStrategy;
+
+public class MetricFactory {
+
+    public static Metric createMetric(MetricType type) {
+        if (type.getType() == 1) {
+            return new NumericMetric<>(type.getName(), new RollingCounter());
+        } else if (type == MetricType.CIRCUIT_BREAKER) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: it's a little confusing to have different cases based on different values - some for `type.getType()` and others for the `type` itself. It might introduce overlaps in the future, and generally it's hard to follow, please either use `type` everywhere, or `getType`.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public enum MetricType {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: so far my impression is you combined the name of the metric and its type in a this enum, which is contradictory to its name. And one consequence of that choice is having the second type field in it, which in my opinion *is* the real type. Another bad side effect is when you instantiate metrics, you get the name of the metric from the type. So I suggest replacing the integer values (for `type` field) with actual enum values, in other words, getting rid of the type field, and making this enum less granular. Have a different utility class where you have constant names for metrics as well as the mapping between the name and the type (many to one is fine).

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import com.amazon.opendistroforelasticsearch.sql.query.join.BackOffRetryStrategy;
+
+public class MetricFactory {
+
+    public static Metric createMetric(MetricType type) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: if you agree to my overall suggestions about refactoring the MetricType class, this method will then accept two parameters: `metricName` and `metricType`.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java (hunk 7 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: Another bad side effect of adding that subtype (`type` field) : no one will understand what those numbers mean, and when one adds a new type with the existing approach s/he won't know which number to assign to it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            registerMetric(MetricFactory.createMetric(metricType));
+        }
+    }
+
+    public void registerMetric(Metric metric) {
+        metricMap.put(metric.getName(), metric);
+    }
+
+    public void unRegisterMetric(String name) {
+        if (name == null || !metricMap.containsKey(name)) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: `Map::remove(key)` checks if the key is present, so this check is redundant here.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            return;
+        }
+
+        metricMap.remove(name);
+    }
+
+    public Metric getMetric(String name) {
+        if (name == null || !metricMap.containsKey(name)) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: Same here, if the key is not present `Map::get` will return `null`, so no need to do the check here.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+public class RollingCounter implements Counter<Long> {
+
+    private final long capacity;
+    private final long window;
+    private final long interval;
+    private final Clock clock;
+    private final ConcurrentSkipListMap<Long, Long> time2CountWin;
+    private final LongAdder count;
+
+    public RollingCounter() {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: please avoid duplicating code:
```java
public RollingCounter() {

  final long window = LocalClusterState.state().getSettingValue(SqlSettings.METRICS_ROLLING_WINDOW);
  final long interval = LocalClusterState.state().getSettingValue(SqlSettings.METRICS_ROLLING_INTERVAL);

  this(window, interval);
}
```

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        this.interval = interval;
+        clock = Clock.systemDefaultZone();
+        time2CountWin = new ConcurrentSkipListMap<>();
+        count = new LongAdder();
+        capacity = window / interval * 2;
+    }
+
+    @Override
+    public void increment() {
+        trim();
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: you can just call `increment(1L)`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java [hunk 13]
```diff
@@ -91,6 +95,11 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
                 return channel -> restExecutor.execute(client, additionalParams, queryAction, channel);
             }
         } catch (Exception e) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java]: ideally this should be
```java
catch (final NullPointerException | SqlParseException | ParserException | SQLFeatureNotSupportedException | IllegalArgumentException | IndexNotFoundException e) {
  // do whatever
} catch (final Exception e) {
  // handle generic case
}
```

but even if you want to keep it this way please update to this:
```java
Metrics.getInstance().getNumericMetric(isClientError(e) ? MetricType.FAILED_REQ_COUNT_CUS : MetricType.FAILED_REQ_COUNT_SYS).increment();
```

Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java]: Also, I know it's not your code, but can you fix `isClientError` method to use `||` instead of `|` .

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class Metrics {
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: Shouldn't this be a singleton?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.elasticsearch.rest.RestStatus.BAD_REQUEST;
+import static org.elasticsearch.rest.RestStatus.SERVICE_UNAVAILABLE;
+
+public class RestSqlStatsAction extends BaseRestHandler {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java]: please do not duplicate code (e.g. methods `responseParams`, `isClientError`, etc.), if needed create another base class with all the shared code, or move the shared code into a util class and include it through containment, anyway you want, but have the common methods in one place only.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java [hunk 16]
```diff
@@ -51,6 +54,8 @@
 
     private final static Object obj = new Object();
 
+    public final static Supplier<Integer> GET_CB_STATE = () -> isMemoryHealthy() ? 0 : 1;
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: what does **CB** stand for?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java [hunk 17]
```diff
@@ -74,12 +79,17 @@ public static boolean isHealthy() {
...
             if (ThreadLocalRandom.current().nextBoolean()) {
+                Metrics.getInstance().getNumericMetric(MetricType.FAILED_REQ_COUNT_CB).increment();
                 LOG.warn("[MCB1] Directly abort on idx {}.", i);
                 return false;
             }
             backOffSleep(intervals[i]);
         }
-        return isMemoryHealthy();
+
+        boolean health = isMemoryHealthy();
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: `boolean isHealthy`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+    public Metric getMetric(String name) {
+        if (name == null || !metricMap.containsKey(name)) {
+            return null;
+        }
+
+        return metricMap.get(name);
+    }
+
+    public NumericMetric getNumericMetric(MetricType metricType) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: np: `getNumericalMetric`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+public interface Counter<T> {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: Why is this class generic? What types other than `long` do you expect to have in the future? It would be a little odd to have a method `increment` for non-integral types, and all integral types can fit under `long`. Do I miss something?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/NumericMetric.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+public class NumericMetric<T> extends Metric<T> {
+
+    private Counter<T> counter;
+
+    public NumericMetric(String name, Counter counter) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/NumericMetric.java]: I think it's not a good idea to "outsource" a concrete type's behavior. What's the use case where we would need to change how we count? And the second comment, even if you want to support this, it should be an override to the default, in other words there should be a constructor which accepts just the name and uses `BasicCounter` as the `Counter` implementation.

Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/NumericMetric.java]: Ok, never mind, I guess this is fine. However, I am thinking if it would be better to derive the Counter implementation from the `MetricType`, I got the impression that those are somehow connected. Anyway, this is an `NP` comment, just out of my curiousity.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java (hunk 13 above)
Comment (dai-chen) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java]: > Also, I know it's not your code, but can you fix `isClientError` method to use `||` instead of `|` .

Good catch. That was my mistake. :)

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java (hunk 1 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: Sure, will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java (hunk 2 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: I consider it as a function, so use all caps.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java (hunk 3 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: Sure, will add some comments.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java (hunk 4 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: I am fine with either.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java (hunk 5 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: Good suggestion, will update.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 6 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: OK, will do it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java (hunk 11 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: I think this won't work, because the constructor call has to be the first, which means you can inline the window and interval calculations.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java (hunk 2 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: Do we have any other code where we use all caps for function names? From what I've seen and used so far, only the names of constants should be in all caps.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java (hunk 12 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: Good suggestions, will do it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java (hunk 7 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricType.java]: Good suggestion, will do it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 8 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: Sure, will do it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class Metrics {
+
+    private static Metrics metrics = new Metrics();
+    private ConcurrentHashMap<String, Metric> metricMap = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: OK, will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java (hunk 9 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: OK, will keep only null check.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java (hunk 10 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: OK, will keep only null check.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java (hunk 11 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: Good suggestion!

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java (hunk 12 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/RollingCounter.java]: Good suggestion!

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java (hunk 13 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java]: Good catch! Thanks.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java (hunk 15 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java]: Will do it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java (hunk 16 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: circuit breaker

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java (hunk 17 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: Will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java (hunk 18 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: OK, will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java (hunk 19 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: Counter only support integer/long increment, but the value could be float etc., for example, error rate.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/NumericMetric.java (hunk 20 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/NumericMetric.java]: Thanks.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java (hunk 2 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: Not sure. I can change it to loadValue, is that OK for you?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java (hunk 19 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: It seems counter-intuitive to me to have a `float` counter, firstly because I wouldn't know what the increment for `float` is. Seems like it should always be `long`, and then the users can always convert it to whatever numerical type.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java (hunk 2 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/GaugeMetric.java]: That is fine, thank you.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.metrics;
+
+import com.amazon.opendistroforelasticsearch.sql.query.join.BackOffRetryStrategy;
+
+public class MetricFactory {
+
+    public static Metric createMetric(MetricName name) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: Should I be able to create more than one `Metric` instance for each `MetricName` ?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        for (MetricName metricName : MetricName.values()) {
+            registerMetric(MetricFactory.createMetric(metricName));
+        }
+    }
+
+    public void registerMetric(Metric metric) {
+        registeredMetricsByName.put(metric.getName(), metric);
+    }
+
+    public void unRegisterMetric(String name) {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: np: `unregisterMetric`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java [hunk 24]
```diff
@@ -74,12 +79,17 @@ public static boolean isHealthy() {
...
+                Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_CB).increment();
                 LOG.warn("[MCB1] Directly abort on idx {}.", i);
                 return false;
             }
             backOffSleep(intervals[i]);
         }
-        return isMemoryHealthy();
+
+        boolean isHealthy = isMemoryHealthy();
+        if (!isHealthy) Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_CB).increment();
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: np: please use proper if block `{ }`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+public enum MetricName {
+
+    REQ_TOTAL("request_total"),
+    REQ_COUNT_TOTAL("request_count"),
+    FAILED_REQ_COUNT_SYS("failed_request_count_syserr"),
+    FAILED_REQ_COUNT_CUS("failed_request_count_cuserr"),
+    FAILED_REQ_COUNT_CB("failed_request_count_cb"),
+    CIRCUIT_BREAKER("circuit_breaker"),
+    DEFAULT("default");
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java]: what is this one for?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+    private static Metrics metrics = new Metrics();
+    private ConcurrentHashMap<String, Metric> registeredMetricsByName = new ConcurrentHashMap<>();
+
+    public static Metrics getInstance() {
+        return metrics;
+    }
+
+    private Metrics() {}
+
+    public void registerDefaultMetrics() {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: how are default metrics different from non-default? If I want to add my own metric, should I add it into the set of default or not?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 22 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: No, Metric is singleton for SQL plugin.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java (hunk 19 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Counter.java]: Not increment for float. I means float for getValue. Anyway, we can make more discussion about it in next release. Thanks.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java (hunk 23 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: Ah, nice, will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java (hunk 24 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/join/BackOffRetryStrategy.java]: OK, will change it.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java (hunk 25 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java]: FAILED_REQ_COUNT_CUS is for counting request failures by customer error, like SQL syntax error etc.
FAILED_REQ_COUNT_CB is for counting request failures by circuit breaker.
CIRCUIT_BREAKER is current status of circuit breaker, i.e. circuit is broken or not.
DEFAULT is a default metric, could be used in some scenarios in the future.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java (hunk 26 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/Metrics.java]: Either is OK, I think. You could add it in this default if you think this metric is should be initialized from beginning. I not, you could choose when and where to register/unregister your own.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 22 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: `Metrics` is, but `Metric` and its subclasses are not.

Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: Oh, I got what you mean. Metric is not singleton, but we only hold one metric per MetricName.

**Comment by abbashus** [comment]
Since the stats will be for a particular node and not whole cluster, we should also add the node-name/node-id to the response. Ex:

```
{
  "node_name" :  "somenode" ,
  "request_count" : 600,
  "failed_request_count_syserr" : 0,
  "failed_request_count_cuserr" : 0,
  "circuit_breaker" : 0 
}
```

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/MetricsIT.java [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.concurrent.TimeUnit;
+
+import static com.amazon.opendistroforelasticsearch.sql.intgtest.TestsConstants.TEST_INDEX_DOG;
```
Comment (qreshi) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/MetricsIT.java]: NP: Since we're using ```esintgtest.TestConstants``` which is in the same directory, this static import isn't necessary. The IDE probably chose the wrong one and added the import.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 22 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: for now yes, but you don't restrict it. I think we should not allow registering metrics with the same name, moreover, not even creating if there is already a registered metric.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/MetricsIT.java (hunk 27 above)
Comment (jngz-es) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/MetricsIT.java]: Make sense, thanks.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java (hunk 25 above)
Comment (arsen-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java]: I think it's better to get rid of default, especially that you have a default case in your `MetricFactory::createMetric` anyway.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java (hunk 22 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricFactory.java]: We could override the metric, just one metric for one name is OK.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestStatus;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.elasticsearch.rest.RestStatus.SERVICE_UNAVAILABLE;
+
+public class RestSqlStatsAction extends BaseRestHandler {
```
Comment (arsen-es) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java]: Let's explicitly document that this API will return the stats for the node that received the request, and that the second change (coming soon) will add support to get cluster level stats.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java (hunk 25 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/metrics/MetricName.java]: I think these two default are two different concepts. One is for user to use intentionally another one is for some unexpected scenarios.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java (hunk 28 above)
Comment (jngz-es) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlStatsAction.java]: Sure, will add it.
